### PR TITLE
Redriven Step Functions Trace Merging

### PR DESCRIPTION
**Pull request by avedmala**
### What does this PR do?

Adds support for Step Functions trace merging in Redrive cases

We previously used `hash(ExecutionId # StateName # StateEnteredTime)` for spanID calculation but these values are identical across redrives for a Lambda task state. The new approach also adds a `RedriveCount` to the end of the hash but omits this value when it is 0 to have easy backwards compatability.

### Motivation

https://github.com/DataDog/logs-backend/pull/84188

### Testing Guidelines

Redriven [Trace](https://dd.datad0g.com/apm/trace/3c21d1b923cf0a6725fd9e4d59c5cf38?graphType=flamegraph&shouldShowLegend=true&sort=time&spanID=5500582247263912696&timeHint=1733340837753) with Lambda and Trace Merging
Screenshot 2024-12-04 at 2 36 08 PM

### Types of Changes

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Misc (docs, refactoring, dependency upgrade, etc.)

### Check all that apply

- [x] This PR's description is comprehensive
- [ ] This PR contains breaking changes that are documented in the description
- [ ] This PR introduces new APIs or parameters that are documented and unlikely to change in the foreseeable future
- [ ] This PR impacts documentation, and it has been updated (or a ticket has been logged)
- [ ] This PR's changes are covered by the automated tests
- [ ] This PR collects user input/sensitive content into Datadog
- [ ] This PR passes the integration tests (ask a Datadog member to run the tests)

### CODE REVIEW

File: src/trace/context/extractors/step-function.spec.ts [hunk 1]
```diff
@@ -42,6 +65,22 @@ describe("StepFunctionEventTraceExtractor", () => {
...
+      StepFunctionContextService.instance(redrivePayload);
+
+      const extractor = new StepFunctionEventTraceExtractor();
+
+      // Payload is sent again for safety in case the instance wasn't previously initialized
+      const traceContext = extractor.extract(redrivePayload);
+      expect(traceContext).not.toBeNull();
+
+      expect(traceContext?.toTraceId()).toBe("1139193989631387307");
+      expect(traceContext?.toSpanId()).toBe("3615526381891637131");
```
Comment (kimi-p) [inline src/trace/context/extractors/step-function.spec.ts:79]: Could you link the PR line in the reducer that matches this test?

Also, is the plan to match traceID only but not spanID?

Comment (avedmala) [reply src/trace/context/extractors/step-function.spec.ts:79]: Here's the [matching line](https://github.com/DataDog/logs-backend/blob/c5ae58dc4f823d2a74b4cba4b01b878cb4e2d72e/domains/serverless/apps/logs-to-traces-reducer/src/main/java/com/dd/logs/serverless/logs_to_traces/reducer/output/stepfunction/trace/spanbuilder/BaseSpanBuilder.java#L173) from logs-backend

Yes the lambda span should be in the same trace but we want the spanID to be different from the original when redrive count was 0

Comment (kimi-p) [reply src/trace/context/extractors/step-function.spec.ts:79]: Can you link the test just to verify that the two matches? Thanks!

**Review by kimi-p — APPROVED** [review]
LGTM

**Pull request merged**